### PR TITLE
Issue#419 Implementation for InstanceAdmin::GetInstance with unit tests for no errors

### PR DESCRIPTION
**Pull request by roopak-qlogic**
PR Created for following:
- Implementing the function itself, Implementing the sample snippet and Implementing the unit tests for "no errors"
Following will be part of Next PR:
- Implementing the integration test, and Implementing the unit tests for "unrecoverable error", "retry with recoverable errors".

I have implemented GetInstance API with another way also as below, but in that I was facing problem, that how to get Instance based on instance ID.

btproto::GetInstanceRequest request;
 request.set_name(instance_id);
//Call RPC call to get response
auto response = ClientUtils::MakeCall(
	*client_, *rpc_policy, *backoff_policy, metadata_update_policy_,
	&InstanceAdminClient::GetInstance, request,
	"InstanceAdmin::GetInstance", status, true);
 if (status.ok()) {
	        result = response;
}

### CODE REVIEW

File: bigtable/client/instance_admin.h [hunk 1]
```diff
@@ -59,6 +59,15 @@ class InstanceAdmin {
    */
   std::vector<google::bigtable::admin::v2::Instance> ListInstances();
 
+  /**
+   * Provide Instance based on Instance ID.
+   *
+   * Return the instance in the project.
+   * @return
```
Comment (coryan) [inline bigtable/client/instance_admin.h]: This whole comment can be summarized as:

```C++
/// Get the details of @p instance_id
```

File: bigtable/client/internal/instance_admin.cc [hunk 2]
```diff
@@ -57,6 +57,39 @@ std::vector<btproto::Instance> InstanceAdmin::ListInstances(
...
+                                             grpc::Status& status) {
+  // Copy the policies in effect for the operation.
+  auto rpc_policy = rpc_retry_policy_->clone();
+  auto backoff_policy = rpc_backoff_policy_->clone();
+
+  // Build the RPC request, try to minimize copying.
+  btproto::Instance result;
+  std::string page_token;
+  do {
+    btproto::ListInstancesRequest request;
```
Comment (coryan) [inline bigtable/client/internal/instance_admin.cc]: This implementation is completely wrong, isn't it?  It is calling `ListInstances`.  Please fix.

File: bigtable/client/internal/instance_admin_test.cc [hunk 3]
```diff
@@ -220,3 +220,21 @@ TEST_F(InstanceAdminTest, ListInstancesUnrecoverableFailures) {
...
   EXPECT_FALSE(status.ok());
 }
+
+/// @test Verify that `bigtable::InstanceAdmin::GetInstance` works in the easy
+/// case.
+TEST_F(InstanceAdminTest, GetInstance) {
+  using namespace ::testing;
+
+  bigtable::noex::InstanceAdmin tested(client_);
+  auto mock_list_instances = create_list_instances_lambda("", "", {"t0", "t1"});
```
Comment (coryan) [inline bigtable/client/internal/instance_admin_test.cc]: And the test is testing for the wrong thing too.

File: bigtable/client/internal/instance_admin.cc [hunk 4]
```diff
@@ -57,6 +57,26 @@ std::vector<btproto::Instance> InstanceAdmin::ListInstances(
...
+  auto rpc_policy = rpc_retry_policy_->clone();
+  auto backoff_policy = rpc_backoff_policy_->clone();
+
+  // Build the RPC request, try to minimize copying.
+  btproto::GetInstanceRequest request;
+  // Set Instance ID
+  request.set_name(project_name_ + "/instances/" + instance_id);
+
+  // Call RPC call to get response
+  auto response = ClientUtils::MakeCall(
```
Comment (coryan) [inline bigtable/client/internal/instance_admin.cc]: You can simplify this even further, just `return ClientUtils::MakeCall(`, right?

File: bigtable/client/internal/instance_admin.h [hunk 5]
```diff
@@ -74,6 +74,19 @@ class InstanceAdmin {
    */
   std::vector<::google::bigtable::admin::v2::Instance> ListInstances(
       grpc::Status& status);
+
+  /**
```
Comment (coryan) [inline bigtable/client/internal/instance_admin.h]: Please remove this comment, we are putting all the doxygen documentation in the non-`noex` class.

File: bigtable/client/internal/instance_admin.cc (hunk 4 above)
Comment (roopak-qlogic) [reply bigtable/client/internal/instance_admin.cc]: yes correct

**Comment by roopak-qlogic** [comment]
Fixing build issue, due to rebase merge conflict issue

File: bigtable/client/internal/instance_admin.cc [hunk 6]
```diff
@@ -57,6 +57,24 @@ std::vector<btproto::Instance> InstanceAdmin::ListInstances(
   return result;
 }
 
+btproto::Instance InstanceAdmin::GetInstance(std::string const& instance_id,
+                                             grpc::Status& status) {
+  // Copy the policies in effect for the operation.
+  auto rpc_policy = rpc_retry_policy_->clone();
+  auto backoff_policy = rpc_backoff_policy_->clone();
+
+  // Build the RPC request, try to minimize copying.
```
Comment (coryan) [inline bigtable/client/internal/instance_admin.cc]: Remote spurious comment.

File: bigtable/client/internal/instance_admin.cc [hunk 7]
```diff
@@ -57,6 +57,24 @@ std::vector<btproto::Instance> InstanceAdmin::ListInstances(
...
+  btproto::GetInstanceRequest request;
+  // Set Instance ID
```
Comment (coryan) [inline bigtable/client/internal/instance_admin.cc]: Remote comment or change to what it actually does (setting the instance **name**).

File: bigtable/client/internal/instance_admin_test.cc [hunk 8]
```diff
@@ -220,3 +234,37 @@ TEST_F(InstanceAdminTest, ListInstancesUnrecoverableFailures) {
...
+
+  // After all the setup, make the actual call we want to test.
+  grpc::Status status;
+  std::string instance_id = "t0";
+  auto actual = tested.GetInstance(instance_id, status);
+  EXPECT_TRUE(status.ok());
+  EXPECT_EQ("projects/the-project/instances/t0", actual.name());
+}
+
+/// @test Verify that `bigtable::InstanceAdmin::GetInstance` works in the fail
```
Comment (coryan) [inline bigtable/client/internal/instance_admin_test.cc]: This is the same as the previous test, right?  Remove it for now.

File: bigtable/client/internal/instance_admin_test.cc [hunk 9]
```diff
@@ -61,6 +61,20 @@ auto create_list_instances_lambda = [](std::string expected_token,
...
 };
 
+// A lambda to create lambdas.  Basically we would be rewriting the same
+// lambda twice without this thing.
+auto create_instance = [](std::string expected_token,
+                          std::string returned_token, std::string instance_id) {
+  return [expected_token, returned_token, instance_id](
+      grpc::ClientContext* ctx, btproto::GetInstanceRequest const& request,
+      btproto::Instance* response) {
+    auto const project_name = "projects/" + kProjectId;
```
Comment (coryan) [inline bigtable/client/internal/instance_admin_test.cc]: I think this is not used, right?  Please remove.

**Pull request merged**